src/utils/fileInspector.js

```diff
@@ -0,0 +1,68 @@
+import { Gunzip, strFromU8 } from 'fflate';
+
+import techOptions from './fileUploadSpecifications';
+import readFileToBuffer from './readFileToBuffer';
+
+const Verdict = {
+  INVALID_NAME: -2,
+  INVALID_FORMAT: -1,
+  VALID_ZIPPED: 0,
+  VALID_UNZIPPED: 1,
+};
+
+const MATRIX_SIGNATURE = Buffer.from('%%MatrixMarket', 'ascii');
+const FEATURES_SIGNATURE = Buffer.from('ENS', 'ascii');
+
+const GZIP_SIGNATURE = Buffer.from([0x1f, 0x8b]);
+
+const inspectFile = async (file, technology) => {
+  // Validate a file requested for upload to the platform.
+
+  // immediately discard file if filename is not in valid set
+  const validNames = techOptions[technology].acceptedFiles;
+  if (!validNames.has(file.name)) {
+    return Verdict.INVALID_NAME;
+  }
+
+  // if name is valid, inspect first 16 bytes to validate format
+  let data = await readFileToBuffer(file.slice(0, 16));
+
+  const isGzipped = !data.slice(0, 2).compare(GZIP_SIGNATURE);
+
+  if (isGzipped) {
+    // if gzipped, decompress a small chunk to further validate contents
+    // eslint-disable-next-line no-unused-vars
+    const gunzip = new Gunzip((chunk, _) => {
+      data = Buffer.from(chunk.slice(0, 16));
+    });
+    gunzip.push(await readFileToBuffer(file.slice(0, 128)));
+  }
+
+  const valid = isGzipped ? Verdict.VALID_ZIPPED : Verdict.VALID_UNZIPPED;
+
+  // check matrix file starts with matrix signature
+  if (file.name.startsWith('matrix')
+    && !data.slice(0, MATRIX_SIGNATURE.length).compare(MATRIX_SIGNATURE)) {
+    return valid;
+  }
+
+  // check genes file starts with Ensembl Stable ID - ENS or "ENS
+  if ((file.name.startsWith('features') || file.name.startsWith('genes'))
+      && (!data.slice(0, 3).compare(FEATURES_SIGNATURE)
+      || !data.slice(1, 4).compare(FEATURES_SIGNATURE))) {
+    return valid;
```

Comment:
Is this going to detect an invalid features file because of this reason here: https://this-is-biomage.slack.com/archives/C0250TDU1PF/p1628172894026400?thread_ts=1628171840.026200&cid=C0250TDU1PF ? @ogibson-biomage will be able to provide you with more details on what exactly was invalid there.

Reply:
Yes, it would catch this! That data seems to have gene names instead of Ensembl IDs.